### PR TITLE
[Target] Fix empty target and host for autotvm task

### PR DESCRIPTION
**Pull request by zxybazh**
In order to patch the empty target and host situation in [Issue 7790](https://github.com/apache/tvm/issues/7790).

### CODE REVIEW

**Review by comaniac — COMMENTED** [review]
Seems to me that `__setstate__` also has this issue? If you save the state with both target and target_host are None, then you will get the same problem in L202.

**Comment by zxybazh** [comment]
Thanks @comaniac for pointing that out. I have added check in this case. In order to prevent this issue, I would take a look at other occurances as well and update in this PR.

**Comment by comaniac** [comment]
> Thanks @comaniac for pointing that out. I have added check in this case. In order to prevent this issue, I would take a look at other occurances as well and update in this PR.

Sure. Please comment here once it's ready for reivew.

**Comment by zxybazh** [comment]
Hi, I've checked all occurences and found other cases possible to reach the same error, e.g., the `_build_module_no_factory` function. Some the the functions calls does not have guarantee about whether target is None. Therefore, instead of adding extra checks in occurrences, I changed the `check_and_update_host_consist` function and add a warning to warn about some corner cases. After all IMO empty target should not be a typical case in the system, and should be avoided in the future. What do you think @comaniac ?

**Comment by zxybazh** [comment]
cc @junrushao1994

File: python/tvm/target/target.py [hunk 1]
```diff
@@ -182,6 +182,10 @@ def check_and_update_host_consist(target, host=None, target_is_dict_key=True):
         target_is_dict_key : Bool
             When the type of target is dict, whether Target is the key (Otherwise the value)
         """
+        if target is None:
+            if host is not None:
+                warnings.warn("Target host is not empty when target is empty.")
```
Comment (comaniac) [inline python/tvm/target/target.py:187]: Should this be a warning or an assertion, if target=None and host!=None doesn't make any sense?

**Review by comaniac — COMMENTED** [review]
Seems like we don't have a test case for `check_and_update_host_consist`. Should we add one?

File: python/tvm/target/target.py (hunk 1 above)
Comment (zxybazh) [reply python/tvm/target/target.py]: Good point, should be an assertion. I fixed it.

**Comment by zxybazh** [comment]
Hi Cody, I don't intend to add test for the function because this is a temporary function for migration to the new target system. It should be deprecated soon after target host deprecation and is not encouraged for any other usage. I have also added the assertion as per your good point. Please review and let me know any other advice, thanks!

File: python/tvm/target/target.py [hunk 2]
```diff
@@ -182,6 +182,9 @@ def check_and_update_host_consist(target, host=None, target_is_dict_key=True):
         target_is_dict_key : Bool
             When the type of target is dict, whether Target is the key (Otherwise the value)
         """
+        if target is None:
+            assert host is None
```
Comment (comaniac) [inline python/tvm/target/target.py:186]: You can keep the message to let people know why it's failed.

**Review by comaniac — COMMENTED** [review]
Well...I don't quite agree on this point. Even this is just a temporary function, it is here now, and it is being used at more than 30 places in the current code base, which means a bug in this function affects 30+ places. Plus, the "tempoaray" could be one day or one to several months. It's not guaranteed that the final PR will be merged by when.

Would like to hear other's opinion.
c @junrushao1994

**Comment by zxybazh** [comment]
Well, it does impact a lot of functions. I have added tests and fixed assertion warning as suggested. Please check, thanks!

**Review by comaniac — APPROVED** [review]
LGTM. Thanks

**Review by junrushao1994 — COMMENTED** [review]
Thanks for the fix!

**Pull request merged**